### PR TITLE
feat: `Fin.foldlM` and `Fin.foldrM`

### PR DESCRIPTION
**Pull request by fgdorais**
- [x] Depends on #827

### CODE REVIEW

**Comment by fgdorais** [comment]
Might need minor adjustments after [lean4#4304](https://github.com/leanprover/lean4/pull/4304)

File: Batteries/Data/Fin/Basic.lean [hunk 1]
```diff
@@ -15,16 +15,67 @@ def enum (n) : Array (Fin n) := Array.ofFn id
...
+  loop ⟨n, Nat.le_refl n⟩ init where
+  /--
+  Inner loop for `Fin.foldrM`.
+  ```
+  Fin.foldrM n f i x₀ = do
+    let x₁ ← f (n-1) x₀
+    let x₂ ← f (n-2) x₁
+    ...
+    let xᵢ ← f (n-i) xᵢ₋₁
+    pure xᵢ
```
Comment (digama0) [inline Batteries/Data/Fin/Basic.lean:66]:
```suggestion
  Fin.foldrM n f i xᵢ = do
    let xᵢ₋₁ ← f (i-1) xᵢ
    ...
    let xᵢ ← f 1 x₂
    let x₀ ← f 0 x₁
    pure x₀
```

File: Batteries/Data/Fin/Basic.lean [hunk 2]
```diff
@@ -15,16 +15,67 @@ def enum (n) : Array (Fin n) := Array.ofFn id
...
+  /--
+  Inner loop for `Fin.foldlM`.
+  ```
+  Fin.foldlM.loop n f xᵢ i = do
+    let xᵢ₊₁ ← f xᵢ i
+    ...
+    let xₙ ← f xₙ₋₁ (n-1)
+    pure xₙ
+  ```
+  `Fin.foldlM.loop n f x i = f x i >>= fun x => f x (i+1) >>= ...`
```
Comment (fgdorais) [inline Batteries/Data/Fin/Basic.lean:39]:
```suggestion
```

**Comment by fgdorais** [comment]
Git cleanup happened. @digama0's docstring correction was finally correctly applied.

**Comment by fgdorais** [comment]
The basic definitions should eventually be moved to core but that can wait until there is less upstreaming pressure.

**Pull request merged**